rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake

```diff
@@ -177,105 +176,94 @@ else()
     ${_generated_msg_py_files}
     ${_generated_msg_c_files}
     ${_generated_srv_files}
+    ${_generated_extension_files}
+    ${_extension_dependencies}
+    ${rosidl_generate_interfaces_TARGET}__rosidl_generator_c
   )
 endif()
 
 macro(set_properties _build_type)
-  set_target_properties(${_msg_name}${_pyext_suffix} PROPERTIES
+  set_target_properties(${_target_name} PROPERTIES
     COMPILE_FLAGS "${_extension_compile_flags}"
     PREFIX ""
-    LIBRARY_OUTPUT_DIRECTORY${_build_type} "${_output_path}/${_parent_folder}"
-    RUNTIME_OUTPUT_DIRECTORY${_build_type} "${_output_path}/${_parent_folder}"
-    OUTPUT_NAME "${_base_msg_name}__${_typesupport_impl}${PythonExtra_EXTENSION_SUFFIX}"
+    LIBRARY_OUTPUT_DIRECTORY${_build_type} ${_output_path}
+    RUNTIME_OUTPUT_DIRECTORY${_build_type} ${_output_path}
+    OUTPUT_NAME "${PROJECT_NAME}_s__${_typesupport_impl}${PythonExtra_EXTENSION_SUFFIX}"
     SUFFIX "${PythonExtra_EXTENSION_EXTENSION}")
 endmacro()
 
-foreach(_generated_msg_c_ts_file ${_generated_msg_c_ts_files})
-  get_filename_component(_full_folder "${_generated_msg_c_ts_file}" DIRECTORY)
-  get_filename_component(_parent_folder "${_full_folder}" NAME)
-  get_filename_component(_base_msg_name "${_generated_msg_c_ts_file}" NAME_WE)
-  get_filename_component(_full_extension_msg_name "${_generated_msg_c_ts_file}" EXT)
+foreach(_typesupport_impl ${_typesupport_impls})
+  find_package(${_typesupport_impl} REQUIRED)
 
-  set(_msg_name "${_base_msg_name}${_full_extension_msg_name}")
-
-  # foreach(_typesupport_impl ${_typesupport_impls})
-    list(FIND _generated_msg_c_ts_files ${_generated_msg_c_ts_file} _file_index)
-    list(GET _type_support_by_generated_msg_c_files ${_file_index} _typesupport_impl)
-    find_package(${_typesupport_impl} REQUIRED)
-    set(_generated_msg_c_common_file "${_full_folder}/${_base_msg_name}.c")
-
-    set(_pyext_suffix "__pyext")
-
-    # TODO(esteve): Change the following code so that each file is compiled independently
-    add_library(${_msg_name}${_pyext_suffix} SHARED
-      "${_generated_msg_c_ts_file}"
-      "${_generated_msg_c_common_file}"
-    )
+  set(_pyext_suffix "__pyext")
+  set(_target_name "${PROJECT_NAME}${_typesupport_impl}${_pyext_suffix}")
```

Comment:
You might want to use a separator between the project name and the type support name (e.g. double underscore).